cpp/src/request_am.cpp

```diff
@@ -5,17 +5,97 @@
 #include <cstdio>
 #include <memory>
 #include <sstream>
+#include <stdexcept>
 #include <string>
 
 #include <ucp/api/ucp.h>
 
+#include <ucs/memory/memory_type.h>
 #include <ucxx/buffer.h>
 #include <ucxx/delayed_submission.h>
 #include <ucxx/internal/request_am.h>
 #include <ucxx/request_am.h>
+#include <ucxx/typedefs.h>
 
 namespace ucxx {
 
+AmReceiverCallbackInfo::AmReceiverCallbackInfo(const AmReceiverCallbackOwnerType owner,
+                                               AmReceiverCallbackIdType id)
+  : owner(owner), id(id)
+{
+}
+
+typedef std::string AmHeaderSerialized;
+
+struct AmHeader {
+  ucs_memory_type_t memoryType;
+  std::optional<AmReceiverCallbackInfo> receiverCallbackInfo;
+
+  static AmHeader deserialize(const AmHeaderSerialized serialized)
+  {
+    size_t offset{0};
+
+    ucs_memory_type_t memoryType;
+    memcpy(&memoryType, serialized.data() + offset, sizeof(memoryType));
+    offset += sizeof(memoryType);
+
+    bool hasReceiverCallback{false};
+    memcpy(&hasReceiverCallback, serialized.data() + offset, sizeof(hasReceiverCallback));
+    offset += sizeof(hasReceiverCallback);
+
+    if (hasReceiverCallback) {
+      size_t ownerSize{0};
+      memcpy(&ownerSize, serialized.data() + offset, sizeof(ownerSize));
+      offset += sizeof(ownerSize);
+
+      auto owner = AmReceiverCallbackOwnerType(ownerSize, 0);
+      memcpy(owner.data(), serialized.data() + offset, ownerSize);
+      offset += ownerSize;
+
+      AmReceiverCallbackIdType id{};
+      memcpy(&id, serialized.data() + offset, sizeof(id));
+      offset += sizeof(id);
+
+      return AmHeader{.memoryType           = memoryType,
+                      .receiverCallbackInfo = AmReceiverCallbackInfo(owner, id)};
+    }
+
+    return AmHeader{.memoryType = memoryType, .receiverCallbackInfo = std::nullopt};
+  }
+
+  const AmHeaderSerialized serialize() const
+  {
+    size_t offset{0};
+    bool hasReceiverCallback{static_cast<bool>(receiverCallbackInfo)};
+    const size_t ownerSize = (receiverCallbackInfo) ? receiverCallbackInfo->owner.size() : 0;
+    const size_t amReceiverCallbackInfoSize =
+      (receiverCallbackInfo) ? sizeof(ownerSize) + ownerSize + sizeof(receiverCallbackInfo->id) : 0;
+    const size_t totalSize =
+      sizeof(memoryType) + sizeof(hasReceiverCallback) + amReceiverCallbackInfoSize;
+    std::string serialized(totalSize, 0);
+
+    memcpy(serialized.data() + offset, &memoryType, sizeof(memoryType));
+    offset += sizeof(memoryType);
```

Comment:
Very minor polish, perhaps unnecessary:
```
auto encode = [&offset, &serialized](void const *data, size_t bytes) {
  memcpy(serialized.data() + offset, data, bytes);
  offset += bytes;
}
```
And then:
```
encode(&memoryType, sizeof(memoryType));
encode(&hasReceiverCallback, sizeof(hasReceiverCallback));
if (hasReceiveCallback) {
  encode(&ownerSize, sizeof(ownerSize));
  encode(receiverCallbackInfo->owner.c_str(), ownerSize);
  encode(&receiverCallbackInfo->id, sizeof(receiverCallbackInfo->id));
}
```

And something matching on the decode side.

Reply:
That's an excellent idea! Done in https://github.com/rapidsai/ucxx/pull/186/commits/1920ff1c6b0ae71599e86bb9303b3d64e46545dc .